lib/Reconciler.lua

```diff
@@ -1,542 +1,361 @@
-				local childInstance = Reconciler._mountInternal(childElement, target, key, context)
-
-				children[key] = childInstance
-			end
-		end
-
-		return {
-			[isInstanceHandle] = true,
-			_key = key,
-			_parent = parent,
-			_element = element,
-			_context = context,
-			_children = children,
-			_rbx = target,
-		}
-	elseif typeof(element) == "boolean" then
-		-- Ignore booleans of either value
-		-- See https://github.com/Roblox/roact/issues/14
-		return nil
 	end
-
-	error(("Cannot mount invalid Roact element %q"):format(tostring(element)))
 end
 
---[[
-	A public interface around _reconcileInternal
-]]
-function Reconciler.reconcile(instanceHandle, newElement)
-	if instanceHandle == nil or not instanceHandle[isInstanceHandle] then
-		local message = (
-			"Bad argument #1 to Reconciler.reconcile, expected component instance handle, found %s"
-		):format(
-			typeof(instanceHandle)
-		)
-
-		error(message, 2)
-	end
+local function taskReconcileNode(details)
+	local node = details.node
+	local toElement = details.toElement
 
-	return Reconciler._reconcileInternal(instanceHandle, newElement)
-end
+	assert(Type.is(node, Type.Node))
+	assert(Type.is(toElement, Type.Element))
 
---[[
-	Applies the state given by newElement to an existing Roact instance.
+	return function(tree)
+		local fromElement = node.element
 
-	reconcile will return the instance that should be used. This instance can
-	be different than the one that was passed in.
-]]
-function Reconciler._reconcileInternal(instanceHandle, newElement)
-	local oldElement = instanceHandle._element
+		-- TODO: Branch on kind of node
+		-- TODO: Check if component type changed
 
-	-- Instance was deleted!
-	if not newElement then
-		Reconciler.unmount(instanceHandle)
+		local visitedProps = {}
 
-		return nil
-	end
+		for prop, newValue in pairs(toElement.props) do
+			visitedProps[prop] = true
+			local oldValue = fromElement.props[prop]
 
-	-- If the element changes type, we assume its subtree will be substantially
-	-- different. This lets us skip comparisons of a large swath of nodes.
-	if oldElement.component ~= newElement.component then
-		local parent = instanceHandle._parent
-		local key = instanceHandle._key
+			DEBUG_print("Checking prop", prop, oldValue, newValue)
 
-		local context
-		if isStatefulElement(oldElement) then
-			context = instanceHandle._instance._context
-		else
-			context = instanceHandle._context
-		end
+			if newValue == oldValue then
+				DEBUG_print("\tProp is the same.")
+			else
+				if prop == Core.Children then
+					DEBUG_print("\tReconciling children...")
 
-		Reconciler.unmount(instanceHandle)
+					for key, newChildElement in pairs(newValue) do
+						local childNode = node.children[key]
 
-		local newInstance = Reconciler._mountInternal(newElement, parent, key, context)
+						local oldChildElement = oldValue[key]
 
-		return newInstance
-	end
+						-- TODO: What if oldValue[Core.Children] is nil?
```

Comment:
That seems like a reasonable idea here. The other alternative would be to drop into the reification routine for children in that case instead of doing comparisons unnecessarily.